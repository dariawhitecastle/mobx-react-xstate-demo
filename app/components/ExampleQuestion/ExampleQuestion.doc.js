import { describe, PropTypes } from 'react-desc'

export const doc = ExampleQuestion => {
  const DocumentedExampleQuestion = describe(ExampleQuestion)
    .description('A ExampleQuestion')
    .usage(
      `import { ExampleQuestion } from 'components/ExampleQuestion';
<ExampleQuestion />`,
    )

  DocumentedExampleQuestion.propTypes = {
    id: PropTypes.string
      .description('The DOM id attribute value to use for the element.')
      .defaultValue(false),
    index: PropTypes.number.isRequired.description('The index of the question component.'),
    goToNextQuestion: PropTypes.func
      .description(
        'The function to call that will increment to the next question in the parent form.',
      )
      .defaultValue(null),
  }

  return DocumentedExampleQuestion
}
