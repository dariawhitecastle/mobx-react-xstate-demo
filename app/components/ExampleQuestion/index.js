export { default as ExampleQuestion } from './ExampleQuestion'
