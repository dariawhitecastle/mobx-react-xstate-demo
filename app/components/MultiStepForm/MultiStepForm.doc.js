import { describe, PropTypes } from 'react-desc'

export const doc = MultiStepForm => {
  const DocumentedMultiStepForm = describe(MultiStepForm)
    .description('A MultiStepForm')
    .usage(
      `import { MultiStepForm } from 'components/MultiStepForm';
<MultiStepForm />`,
    )

  DocumentedMultiStepForm.propTypes = {
    questions: PropTypes.array.isRequired.description(
      'An array of questions to be cycled through within the form.',
    ),
  }

  return DocumentedMultiStepForm
}
