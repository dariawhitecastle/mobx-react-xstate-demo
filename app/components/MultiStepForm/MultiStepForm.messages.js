/*
 * MultiStepForm Messages
 *
 * This contains all the text for the MultiStepForm component.
 */

import { defineMessages } from 'react-intl'

export const scope = 'app.components.MultiStepForm'

export default defineMessages({
  header: {
    id: `${scope}.header`,
    defaultMessage: 'This is the MultiStepForm component!',
  },
})
