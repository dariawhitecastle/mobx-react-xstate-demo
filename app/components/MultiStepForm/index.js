export { default as MultiStepForm } from './MultiStepForm'
