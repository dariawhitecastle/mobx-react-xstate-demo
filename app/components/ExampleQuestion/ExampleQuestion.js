/* eslint-disable react/prop-types */
/**
 * ExampleQuestion
 */

import React, { useContext } from 'react'
import { observer } from 'mobx-react'

// Utils
import { doc } from './ExampleQuestion.doc'
import helpers from '../../utils/helpers'

// Components
import { Button } from '../Button'

// Stores
import { OnboardingStoreContext } from '../../stores/OnboardingStore'

const ExampleQuestion = observer(props => {
  const onboardingStore = useContext(OnboardingStoreContext)
  onboardingStore.currentQuestionIndex = props.index

  /**
   * Update the onboarding form responses in the OnboardingStore.
   */
  const updateOnboardingForm = () => {
    const { goToNextQuestion } = props

    onboardingStore.addResponse({
      exampleKey: 'exampleValue',
    })

    goToNextQuestion()
  }

  return (
    <div>
      <p>{props.index}</p>
      <Button onClick={updateOnboardingForm} />
    </div>
  )
})

export default helpers.createWithDoc({
  docFunction: doc,
  component: ExampleQuestion,
})
