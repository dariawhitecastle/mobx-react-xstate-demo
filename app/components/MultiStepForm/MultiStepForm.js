/* eslint-disable react/prop-types */
import React, { useState } from 'react'
import { CSSTransitionGroup } from 'react-transition-group'

import helpers from '../../utils/helpers'
import { doc } from './MultiStepForm.doc'

/**
 * MultiStepForm
 *
 */
const MultiStepForm = props => {
  const [currentIndex, setCurrentIndex] = useState(props.initialIndex || 0)
  const [transitionName, setTransitionName] = useState('slide')
  /**
   * Increment the current index of the form. Causes `renderQuestion` to
   * increment to the next question.
   *
   * Check to see if the question length - 1 is greater than the currentIndex
   * before incrementing. We don't want to accidentally increment the index
   * without having enough questions.
   */
  const goToNextQuestion = () => {
    const { questions } = props
    setTransitionName('slide')

    if (questions.length - 1 > currentIndex) {
      const newIndex = currentIndex + 1
      setCurrentIndex(newIndex)
    } else {
      console.info('Already at the last question, cannot increment index.')
    }
  }

  const goToPrevQuestion = () => {
    if (!currentIndex) {
      console.info('Cannot decrement index, as it will result in negative value')
    }
    const newIndex = currentIndex - 1
    setCurrentIndex(newIndex)
    setTransitionName('slide-back')
  }

  /**
   * Renders the component for the current index. Passes the `goToNextQuestion`
   * function down to be used by the specified question component.
   */
  const renderQuestion = () => {
    const { questions, ...rest } = props
    const Question = questions[currentIndex]

    return (
      <Question
        key={currentIndex}
        goToNextQuestion={goToNextQuestion}
        goToPrevQuestion={goToPrevQuestion}
        index={currentIndex}
        {...rest}
      />
    )
  }

  return (
    <CSSTransitionGroup
      transitionName={transitionName}
      transitionEnterTimeout={800}
      transitionLeaveTimeout={500}
    >
      {renderQuestion()}
    </CSSTransitionGroup>
  )
}

export default helpers.createWithDoc({
  docFunction: doc,
  component: MultiStepForm,
})
