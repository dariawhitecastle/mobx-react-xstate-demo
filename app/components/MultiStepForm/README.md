<!-- ! This is a generated file. To make changes, edit <Component>.doc.js ! -->
## MultiStepForm
A MultiStepForm

## Usage

```javascript
import { MultiStepForm } from 'components/MultiStepForm';
<MultiStepForm />
```

## Properties

**questions**

Required. An array of questions to be cycled through within the form.

```
array
```
  