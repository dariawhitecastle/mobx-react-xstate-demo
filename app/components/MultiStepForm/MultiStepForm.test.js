/**
 *
 * Tests for MultiStepForm
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react'
import { render } from '@testing-library/react'
import { IntlProvider } from 'react-intl'

import MultiStepForm from './MultiStepForm'
import { ExampleQuestion } from '../ExampleQuestion'
import { DEFAULT_LOCALE } from '../../i18n'

const renderComponent = (props = { questions: [ExampleQuestion, ExampleQuestion] }) =>
  render(
    <IntlProvider locale={DEFAULT_LOCALE}>
      <MultiStepForm {...props} />
    </IntlProvider>,
  )

test('it renders and matches snapshot', () => {
  const { container } = renderComponent()
  expect(container).toMatchSnapshot()
})

test('It does not log to the console', () => {
  const spy = jest.spyOn(global.console, 'log')
  render(
    <IntlProvider locale={DEFAULT_LOCALE}>
      <MultiStepForm questions={[ExampleQuestion, ExampleQuestion]} />
    </IntlProvider>,
  )
  expect(spy).not.toHaveBeenCalled()
})
