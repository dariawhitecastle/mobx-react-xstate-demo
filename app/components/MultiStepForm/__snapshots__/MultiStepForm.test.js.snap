// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches snapshot 1`] = `
<div>
  <span>
    <div>
      <p>
        0
      </p>
      <button
        class="StyledButton-sc-323bzc-0 kHrfdt"
        type="button"
      />
    </div>
  </span>
</div>
`;
